gnosis/safe-react
# Migrate localStorage from subdomains to the main domain

## What it solves
Resolves #2632

## How this PR fixes it
When the app is loaded, if the "parent hostname" is correct:

![image](https://user-images.githubusercontent.com/20442784/135283351-5afb1994-c81f-471e-8b00-4f2cb35ec4c4.png)

it creates an iframe and loads every "child hostname" subdomain:
![image](https://user-images.githubusercontent.com/20442784/135284771-fa15afdc-4ef2-45bd-81c0-47a9d49f0cd1.png)

querying the localStorage of each for data. All of which are sent via `postMessage()` to the "parent hostname".

- The address book is merged -- if there is a repeated entry (checked via `address` and `chainId` match), the name will be replaced.
- All other localStorage values are replaced. We only have mainnet on our main domain, so all other network values will be added.

After migration has occurred, a localStorage flag `SAFE__migratedSubdomains` is set to prevent any future attempts at migration.

## How to test it
1. Initialise the localStorage values here:
  - Polygon: https://pr2778--safereact.review.gnosisdev.com/polygon/app
2. Once initialised, open the unified app here:
  - Mainnet https://pr2695--safereact.review-safe.gnosisdev.com/app/mainnet
3. The data should be migrated automatically and flag set in localStorage.
4. To repeat the migration process, the flag needs to be deleted. You should also clear the address book/migrated networks if you would like to populate the data from subdomain again.
